trainer: User guide for PyTorch Training (#4053)
kubeflow/website: content/en/docs/components/trainer/user-guides/pytorch.md

- `os.environ["LOCAL_RANK"]` - Rank of the current process within a single PyTorch training node.

You can use these values to, for example, download the dataset only on the node with `local_rank=0`,
or export your fine-tuned LLM only on the node with `rank=0` (e.g., the master node).

You can access the distributed environment as follows:

```py
from kubeflow.trainer import TrainerClient, CustomTrainer

def get_torch_dist():
    import os
    import torch
    import torch.distributed as dist

    device, backend = ("cuda", "nccl") if torch.cuda.is_available() else ("cpu", "gloo")
    dist.init_process_group(backend=backend)

    print("PyTorch Distributed Environment")
    print(f"Using device: {device}")
    print(f"WORLD_SIZE: {dist.get_world_size()}")
    print(f"RANK: {dist.get_rank()}")
    print(f"LOCAL_RANK: {os.environ['LOCAL_RANK']}")

job_id = TrainerClient().train(
    runtime=TrainerClient().get_runtime("torch-distributed"),
    trainer=CustomTrainer(
        func=get_torch_dist,
        num_nodes=3,
        resources_per_node={
            "cpu": 2,
        },
    ),
)

while True:
    if TrainerClient().get_job(name=job_id).status == "Succeeded":
        break

>     time.sleep(1)

Reviewer: Same as above comments.

Reviewer 2: Same comment as here: https://github.com/kubeflow/website/pull/4053#discussion_r2221910420, this API will return nothing until Pods are created by TrainJob controller.
What do we think about it ?

across multiple GPUs. The gradients are calculated locally on each GPU and then synchronized globally
to update the model parameters.

In FSDP training, both the dataset and the model gets chopped into slices and assigned to the
different GPUs. The model is split into shards, each hosted on a different GPU. Gradients and
parameter updates are computed locally and synchronized globally.
FSDP is particularly useful for training very large models that cannot fit into the memory of a single GPU.

## Get PyTorch Runtime Packages

Kubeflow Trainer includes a PyTorch runtime called `torch-distributed`, which comes with several
pre-installed Python packages.

Run the following command to get a list of the available packages:

```py
# TODO(andreyvelich): This should be changed to the `get_runtime_packages()` API.
from kubeflow.trainer import TrainerClient, Runtime, CustomTrainer
import time

job_id = TrainerClient().train(
    runtime=TrainerClient().get_runtime("torch-distributed"),
)

while True:
    if TrainerClient().get_job(name=job_id).status == "Succeeded":
        break
    time.sleep(1)

> print(TrainerClient().get_job_logs(name=job_id)["node-0"])

Reviewer 2: I can remove the status check, however that code snippet just returns nothing in that case.
The `get_job_logs()` API exists the loop if Pods haven't been created for TrainJob yet: https://github.com/kubeflow/sdk/blob/main/python/kubeflow/trainer/api/trainer_client.py#L353-L354

> print(TrainerClient().get_job_logs(name=job_id)["node-0"])

Reviewer: In that case, shall we have a new API named `wait_for_job_conditions` like v1?

https://github.com/kubeflow/trainer/blob/52e41d5d124be36d7eb9754e47f7b21efe67b1ac/sdk/python/kubeflow/training/api/training_client.py#L992

You can use these values to, for example, download the dataset only on the node with `local_rank=0`,
or export your fine-tuned LLM only on the node with `rank=0` (e.g., the master node).
nly on the node with local_rank=0, or export your fine-tuned LLM only on the node with rank=0 (e.g., the master node).

You can access the distributed environment as follows:

```py
from kubeflow.trainer import TrainerClient, CustomTrainer

def get_torch_dist():
    import os
    import torch
    import torch.distributed as dist

    device, backend = ("cuda", "nccl") if torch.cuda.is_available() else ("cpu", "gloo")
    dist.init_process_group(backend=backend)

    print("PyTorch Distributed Environment")
    print(f"Using device: {device}")
    print(f"WORLD_SIZE: {dist.get_world_size()}")
    print(f"RANK: {dist.get_rank()}")
    print(f"LOCAL_RANK: {os.environ['LOCAL_RANK']}")

job_id = TrainerClient().train(
    runtime=TrainerClient().get_runtime("torch-distributed"),
    trainer=CustomTrainer(
        func=get_torch_dist,
        num_nodes=3,
        resources_per_node={
            "cpu": 2,
        },
    ),
)

> # TODO (andreyvelich): Change it to is_job_complete() API.

Reviewer 2: I just leave it for us to not forget to update this guide, but I can remove it for now.

    )
    # Configure the pre-trained model and tokenizer.
    tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-32B")
    model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen3-32B")

    # Configure the PyTorch training loop.
    for epoch in range(10):
        for batch_idx, batch in enumerate(train_loader):
            output = model(...)
            model.backward(output.loss)
            model.step()
            ...
    if dist.get_rank() == 0:
        # Export your model to the object storage (e.g. S3)
        boto3.upload_file()
```

### Create TrainJob with `train()` API

After configuring the PyTorch training function, use the `train()` API to create TrainJob:

```python
job_id = TrainerClient().train(
    runtime=TrainerClient().get_runtime("torch-distributed"),
    trainer=CustomTrainer(
        func=fine_tune_qwen,
        num_nodes=2,
        resources_per_node={
            "gpu": 4
        },
        # These packages will be installed on every training node.
        packages_to_install=["transformers>=4.53.0", "boto3"],
    )
)
```

### Get the TrainJob Results

You can use the `get_job_logs()` API to see your TrainJob logs:

```py

> print(TrainerClient().get_job_logs(name=job_id)["node-0"])

Reviewer 3: Ah right, maybe pretty printer would work:

```
from pprint import pprint
pprint.pprint(TrainerClient().get_job_logs(name=job_id, node_rank=0)
```

> print(TrainerClient().get_job_logs(name=job_id)["node-0"])

Reviewer 3: Thanks, it's not that "pretty" after all. Maybe the original code is just as good.

weight = 10
+++

> This page describes `PyTorchJob` for training a machine learning model with [PyTorch](https://pytorch.org/).

Reviewer 4: ```suggestion
This page describes how to use `PyTorchJob` for training a machine learning model with [PyTorch](https://pytorch.org/).
```

> This page describes how to use `PyTorchJob` for training a machine learning model with [PyTorch](https://pytorch.org/).

Reviewer 2: ```suggestion
This page describes how to use `TrainJob` for training a machine learning model with [PyTorch](https://pytorch.org/).
```

The `PyTorchJob` is a Kubernetes
[custom resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
to run PyTorch training jobs on Kubernetes. The Kubeflow implementation of
the `PyTorchJob` is in the [`trainer`](https://github.com/kubeflow/trainer).

> ## Creating a PyTorch Training Job

Reviewer 2: @franciscojavierarceo @kubeflow/wg-training-leads @izuku-sds @astefanutti @Electronic-Waste @varodrig @StefanoFioravanzo @hbelmiro Do we need some sort of introduction to ML Users before this guide ?
For example, what is Kubeflow Trainer Runtime, what is Kubeflow TrainJob from the SDK perspective. 
Maybe we should explain the structure of[ these SDK objects](https://github.com/kubeflow/trainer/blob/master/sdk/kubeflow/trainer/types/types.py#L78-L81) for the ML Users, before we can walk them through Job submission.

I am looking for your suggestions!

> ### Define the Training Function

Reviewer 2: We need to figure out how to avoid duplications with this Getting Started section: https://www.kubeflow.org/docs/components/trainer/getting-started/#getting-started-with-pytorch

> The first step is to create function to train CNN model using Fashion MNIST data. Below is an example of a PyTorch training function using Distributed Data Parallel (DDP). The function configures distributed training, downloads the FashionMNIST dataset (only on the primary node), and runs training loops across workers. The code also shows how to clean up the distributed group at the end.

Reviewer 4: ```suggestion
The first step is to create a function to train a CNN model using the Fashion MNIST dataset. Below is an example of a PyTorch training function using Distributed Data Parallel (DDP). The function configures distributed training, downloads the FashionMNIST dataset (only on the primary node), and runs training loops across workers. The code also shows how to clean up the distributed group at the end.
```

        # Iterate over mini-batches from the training set
        for batch_idx, (inputs, labels) in enumerate(train_loader):
            # Copy the data to the GPU device if available
            inputs, labels = inputs.to(device), labels.to(device)
            # Forward pass
            outputs = model(inputs)
            loss = F.nll_loss(outputs, labels)
            # Backward pass
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            if batch_idx % 10 == 0 and dist.get_rank() == 0:
                print(
                    "Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}".format(
                        epoch,
                        batch_idx * len(inputs),
                        len(train_loader.dataset),
                        100.0 * batch_idx / len(train_loader),
                        loss.item(),
                    )
                )

    # Wait for the distributed training to complete
    dist.barrier()
    if dist.get_rank() == 0:
        print("Training is finished")

    # Finally clean up PyTorch distributed
    dist.destroy_process_group()
```

### Scale PyTorch DDP with Kubeflow TrainJob

> You can use `TrainerClient()` from the Kubeflow SDK to communicate with Kubeflow Trainer APIs and scale your training function across multiple PyTorch training nodes.

Reviewer 4: ```suggestion
You can use the `TrainerClient()` from the Kubeflow SDK to communicate with Kubeflow Trainer APIs and scale your training function across multiple PyTorch training nodes.
```

> `TrainerClient()` verifies that you have required access to the Kubernetes cluster.

Reviewer 4: ```suggestion
`TrainerClient()` verifies that you have the required access to the Kubernetes cluster.
```

> Kubeflow Trainer creates a `TrainJob` resource and automatically sets the appropriate environment variables to set up PyTorch in distributed environment.

Reviewer 4: ```suggestion
The Kubeflow Trainer creates a `TrainJob` resource and automatically sets the appropriate environment variables to set up PyTorch in a distributed environment.
```

```python
from kubeflow.trainer import CustomTrainer, TrainerClient

client = TrainerClient()
```

### List Available Training Runtimes

You can get the list of available Training Runtimes to start your TrainJob.

```python
torch_runtime = None
for runtime in client.list_runtimes():
    print(runtime)
    if runtime.name == "torch-distributed":
        torch_runtime = runtime
```

> Additionally, it might show available accelerator type and number of available resources.

Reviewer 4: ```suggestion
Additionally, it might show available accelerator types and the number of available resources.
```

```bash
Runtime(name='mpi-distributed', trainer=Trainer(trainer_type=<TrainerType.CUSTOM_TRAINER: 'CustomTrainer'>, framework=<Framework.TORCH: 'torch'>, entrypoint=['torchrun'], accelerator='Unknown', accelerator_count=1), pretrained_model=None)
Runtime(name='torch-distributed', trainer=Trainer(trainer_type=<TrainerType.CUSTOM_TRAINER: 'CustomTrainer'>, framework=<Framework.TORCH: 'torch'>, entrypoint=['torchrun'], accelerator='Unknown', accelerator_count='Unknown'), pretrained_model=None)
```

> ### Configure train() API

Reviewer 4: ```suggestion
### Configure the train() API
```

> Kubeflow TrainJob will train the above model on 2 PyTorch nodes. You can specify the resources needed per node.

Reviewer 4: ```suggestion
The Kubeflow TrainJob will train the above model on 2 PyTorch nodes. You can specify the resources needed per node.
```

```python
job_name = client.train(
    trainer=CustomTrainer(
        func=train_pytorch,
        # Set how many PyTorch nodes you want to use for distributed training.
        num_nodes=2,
        # Set the resources for each PyTorch node.
        resources_per_node={
            "cpu": 2,
            "memory": "1Gi",
            # Uncomment this to distribute the TrainJob using GPU nodes.
            # "nvidia.com/gpu": 1,
        },
    ),
    runtime=torch_runtime,
)
```

---

## Monitoring and Managing a PyTorchJob
### Check TrainJob steps

Since the TrainJob performs distributed training across 2 nodes, it generates 2 steps: `node-0`, `node-1`.
```python
import time

def wait_for_job_running():
    for _ in range(100):
        trainjob = client.get_job(name=job_name)
        for step in trainjob.steps:
            if step.name == "node-0" and step.status == "Running":
                return
        print("Wait for TrainJob running status. Sleep for 5 seconds")
        time.sleep(5)

wait_for_job_running()
```

> After nodes are in running state, you can get individual status of each step.

Reviewer 4: ```suggestion
After nodes are in the running state, you can get the individual status of each step.
```

[node-0]: Train Epoch: 1 [11000/60000 (73%)]	Loss: 0.544151
[node-0]: Train Epoch: 1 [12000/60000 (80%)]	Loss: 0.473595
[node-0]: Train Epoch: 1 [13000/60000 (87%)]	Loss: 0.491940
[node-0]: Train Epoch: 1 [14000/60000 (93%)]	Loss: 0.507782
[node-0]: Train Epoch: 2 [0/60000 (0%)]	Loss: 0.504748
[node-0]: Train Epoch: 2 [1000/60000 (7%)]	Loss: 0.448021
[node-0]: Train Epoch: 2 [2000/60000 (13%)]	Loss: 0.438320
[node-0]: Train Epoch: 2 [3000/60000 (20%)]	Loss: 0.308895
[node-0]: Train Epoch: 2 [4000/60000 (27%)]	Loss: 0.421823
[node-0]: Train Epoch: 2 [5000/60000 (33%)]	Loss: 0.397320
[node-0]: Train Epoch: 2 [6000/60000 (40%)]	Loss: 0.521149
[node-0]: Train Epoch: 2 [7000/60000 (47%)]	Loss: 0.505759
[node-0]: Train Epoch: 2 [8000/60000 (53%)]	Loss: 0.334409
[node-0]: Train Epoch: 2 [9000/60000 (60%)]	Loss: 0.381091
[node-0]: Train Epoch: 2 [10000/60000 (67%)]	Loss: 0.278955
[node-0]: Train Epoch: 2 [11000/60000 (73%)]	Loss: 0.409761
[node-0]: Train Epoch: 2 [12000/60000 (80%)]	Loss: 0.337577
[node-0]: Train Epoch: 2 [13000/60000 (87%)]	Loss: 0.390389
[node-0]: Train Epoch: 2 [14000/60000 (93%)]	Loss: 0.288414
[node-0]: Training is finished
```
</details>

### Delete the TrainJob

> When TrainJob is finished, you can delete the resource.

Reviewer 4: ```suggestion
When the TrainJob is finished, you can delete the resource.
```
